### PR TITLE
feat(DTFS2-7120): add audit logs to gef routes of portal api

### PR DESCRIPTION
**Pull request by MarRobSoftwire**
## Introduction :pencil2:
All endpoints in portal-api should add `auditRecords` when inserting/updating data. This PR covers the gef router.
## Resolution :heavy_check_mark:
- Add audit record to database changes for gef routes

### CODE REVIEW

File: portal-api/api-tests/v1/gef/eligibility-criteria.api-test.js [hunk 1]
```diff
@@ -105,6 +106,7 @@ describe(baseUrl, () => {
         isInDraft: expect.any(Boolean),
         createdAt: expect.any(Number),
         criteria: expect.any(Array),
+        auditRecord: generateParsedMockPortalUserAuditDatabaseRecord(anAdmin._id),
```
Comment (MarRobSoftwire) [inline portal-api/api-tests/v1/gef/eligibility-criteria.api-test.js:109]: This test is currently failing. It looks like this submission adds a second `version 2.1` EC to the database( Mock data loader has already added this). When the GET request is made, it returns the first one (i.e. that made my MDL) so the test was passing previously, but not anymore

File: portal-api/api-tests/v1/gef/application.api-test.js [hunk 2]
```diff
@@ -39,6 +40,11 @@ const mockSuccessfulResponse = {
   },
 };
 
+const expectedEligibilityCriteriaAuditRecord = {
+  ...generateParsedMockPortalUserAuditDatabaseRecord('abcdef123456abcdef123456'),
+  lastUpdatedByPortalUserId: expect.any(String),
```
Comment (MarRobSoftwire) [inline portal-api/api-tests/v1/gef/application.api-test.js:45]: These ECs are uploaded by Mock Data Loader so we dont have access to the portal user id.

File: portal-api/src/v1/gef/controllers/clone-gef-deal.controller.js [hunk 3]
```diff
@@ -161,7 +164,7 @@ exports.clone = async (req, res) => {
   if (response.status === 200) {
     const { newDealId } = response;
     // clone the corresponding facilities
-    await cloneFacilities(existingDealId, newDealId);
+    await cloneFacilities(existingDealId, newDealId, req.user);
```
Comment (AlexBramhill) [inline portal-api/src/v1/gef/controllers/clone-gef-deal.controller.js:167]: sorry to be a pain but can we create the audit details inside `exports.clone` and pass it in please?

Comment (AlexBramhill) [reply portal-api/src/v1/gef/controllers/clone-gef-deal.controller.js:167]: (same with cloneDeals)

File: portal-api/src/v1/gef/controllers/application-submit.js [hunk 4]
```diff
@@ -69,7 +69,7 @@ const generateUkefId = async (entity, application) => {
   }
 };
 
-const addSubmissionDateToIssuedFacilities = async (dealId) => {
+const addSubmissionDateToIssuedFacilities = async (dealId, sessionUser) => {
```
Comment (AlexBramhill) [inline portal-api/src/v1/gef/controllers/application-submit.js:72]: any reason why were not doing this at the upper most level please?

**Review by AlexBramhill — COMMENTED** [review]
looking good

File: portal-api/src/v1/gef/controllers/application-submit.js [hunk 5]
```diff
@@ -182,10 +183,11 @@ const submissionPortalActivity = async (application) => {
 
 /**
  * Check the `coverDateConfirmed` property of the facility has the correct boolean flag.
- * @param {Object} app Application object
- * @returns {Bool} Facility(ies) was(were) updated or not
+ * @param {object} app Application object
+ * @param {import("@ukef/dtfs2-common/src/types/audit-details").AuditDetails} auditDetails - user making the request
+ * @returns {boolean} Facility(ies) was(were) updated or not
```
Comment (AlexBramhill) [inline portal-api/src/v1/gef/controllers/application-submit.js:188]:
```suggestion
 * @returns {Promise<Boolean>} Facility(ies) was(were) updated or not
```

File: portal-api/src/v1/gef/controllers/application.controller.js [hunk 6]
```diff
@@ -246,11 +249,12 @@ exports.changeStatus = async (req, res) => {
   let applicationUpdate = { status, ...{ updatedAt: Date.now() } };
 
   if (status === DEAL_STATUS.SUBMITTED_TO_UKEF) {
-    const submissionData = await addSubmissionData(dealId, existingApplication);
+    const submissionData = await addSubmissionData(dealId, existingApplication, generatePortalAuditDetails(req.user._id));
 
     applicationUpdate = {
       ...applicationUpdate,
       ...submissionData,
+      auditRecord: generatePortalUserAuditDatabaseRecord(req.user._id),
```
Comment (AlexBramhill) [inline portal-api/src/v1/gef/controllers/application.controller.js:257]: Sorry big pain thing -- if audit details change later, we might miss updating here. To solve this, could we do the below please?

```suggestion
    const auditDetails = generatePortalAuditDetails(req.user._id);
    ...
    const submissionData = await addSubmissionData(dealId, existingApplication, auditDetails);

    applicationUpdate = {
      ...applicationUpdate,
      ...submissionData,
      auditRecord: generatePortalUserAuditDatabaseRecord(auditDetails),
```

**Review by AlexBramhill — COMMENTED** [review]
soz two small comments but rest is great

**Pull request merged**